### PR TITLE
fix the issue that Heron ECO stream builder does not handle IStatefulWindowedBolt

### PR DESCRIPTION
**Pull request by simingweng**
- fix style check error
- make Heron ECO stream builder handle IStatefulWindowedBolt
- add unit test

@joshfischer1108 @nwangtw

### CODE REVIEW

File: tools/java/src/org/apache/bazel/cppcheck/CppCheck.java [hunk 1]
```diff
@@ -113,7 +114,7 @@ private static void runChecker(List<String> command) throws IOException {
 
     if (cppcheck.exitValue() == 1) {
       LOG.warning("cppcheck detected bad cpp files.");
-      System.exit(1);
+      throw new RuntimeException("cppcheck detected bad cpp files.");
```
Comment (joshfischer1108) [inline tools/java/src/org/apache/bazel/cppcheck/CppCheck.java]: Just curious.  What made you add this in for the PR?

File: eco/src/java/org/apache/heron/eco/builder/heron/StreamBuilder.java [hunk 2]
```diff
@@ -68,6 +70,15 @@ protected void buildStreams(EcoExecutionContext executionContext, TopologyBuilde
               topologyDefinition.parallelismForBolt(stream.getTo()));
           declarers.put(stream.getTo(), declarer);
         }
+      } else if (boltObj instanceof IStatefulWindowedBolt) {
```
Comment (joshfischer1108) [inline eco/src/java/org/apache/heron/eco/builder/heron/StreamBuilder.java]: LGTM

**Comment by simingweng** [comment]
The issue still persist even with this fix in our environment. There looks more work to troubleshoot.

File: tools/java/src/org/apache/bazel/cppcheck/CppCheck.java (hunk 1 above)
Comment (simingweng) [reply tools/java/src/org/apache/bazel/cppcheck/CppCheck.java]: This is to fix a code styling issue that prevents the Heron docker image from being built.

**Comment by nwangtw** [comment]
lgtm

**Comment by nwangtw** [comment]
@simingweng The build error is from here:

Starting audit...
[ERROR] /home/travis/.cache/bazel/_bazel_travis/be6dac4936703c7eedcb4f5cf38cdd65/execroot/org_apache_heron/eco/src/java/org/apache/heron/eco/builder/heron/StreamBuilder.java:47: Line is longer than 100 characters (found 141). [LineLength]
Audit done.
Checkstyle ends with 1 errors.

**Comment by simingweng** [comment]
@nwangtw as I mentioned in the previous comment. The code change in this pull request does not completely solve the problem. Even with the fix, the `initState()` and `preSave()` callback are still not invoked for a stateful bolt defined through ECO. But, I do see in the log that the bolt receives checkpoint request from Topology Master. I’ll need to look deeper, but kinda of stuck now. @joshfischer1108 any clue?

**Comment by nwangtw** [comment]
Hmm. Interesting. Calling setBolt() with IStatefulWindowedBolt should create a StatefulWindowedBoltExecutor object, which should call the initState() and preSave() functions.

**Comment by joshfischer1108** [comment]
I'll have to double check the code.. I wonder if somewhere along the line the bolt is being casted to an unexpected object.

**Comment by joshfischer1108** [comment]
@simingweng  I've re-looked over the code. I don't see any reason why adding the code you wrote wouldn't work.  Can you log which type of bolts are being returned in the topology builder somehow before submission?

**Comment by nwangtw** [comment]
> @simingweng @nwangtw I'll have to double check the code.. I wonder if somewhere along the line the bolt is being casted to an unexpected object.

Yeah. Code looks fine. There ain't many classes along the path but it must have been changed in some way. Otherwise it should work.

**Comment by simingweng** [comment]
One of the integration test on a sliding time window bolt has failed the CI. It’s not a stateful topology, isn’t it? I would assume it’s irrelevant to the change in this PR. @nwangtw @joshfischer1108 ideas?

`[2019-04-18 23:26:22 +0000] [ERROR]: Actual result did not match expected result
[2019-04-18 23:26:22 +0000] [INFO]: Actual result ---------- 
['{"1":[{"tuplesInWindow":["1"]},{"newTuples":["1"]},{"expiredTuples":[]}]}', '{"2":[{"tuplesInWindow":["2","3"]},{"newTuples":["2","3"]},{"expiredTuples":["1"]}]}', '{"3":[{"tuplesInWindow":["4"]},{"newTuples":["4"]},{"expiredTuples":["2","3"]}]}', '{"4":[{"tuplesInWindow":["5"]},{"newTuples":["5"]},{"expiredTuples":["4"]}]}', '{"5":[{"tuplesInWindow":["6"]},{"newTuples":["6"]},{"expiredTuples":["5"]}]}', '{"6":[{"tuplesInWindow":["7"]},{"newTuples":["7"]},{"expiredTuples":["6"]}]}', '{"7":[{"tuplesInWindow":["8"]},{"newTuples":["8"]},{"expiredTuples":["7"]}]}', '{"8":[{"tuplesInWindow":["9"]},{"newTuples":["9"]},{"expiredTuples":["8"]}]}', '{"9":[{"tuplesInWindow":["10"]},{"newTuples":["10"]},{"expiredTuples":["9"]}]}']
[2019-04-18 23:26:22 +0000] [INFO]: Expected result ---------- 
['{"1":[{"tuplesInWindow":["1"]},{"newTuples":["1"]},{"expiredTuples":[]}]}', '{"10":[{"tuplesInWindow":["10"]},{"newTuples":["10"]},{"expiredTuples":["9"]}]}', '{"2":[{"tuplesInWindow":["2"]},{"newTuples":["2"]},{"expiredTuples":["1"]}]}', '{"3":[{"tuplesInWindow":["3"]},{"newTuples":["3"]},{"expiredTuples":["2"]}]}', '{"4":[{"tuplesInWindow":["4"]},{"newTuples":["4"]},{"expiredTuples":["3"]}]}', '{"5":[{"tuplesInWindow":["5"]},{"newTuples":["5"]},{"expiredTuples":["4"]}]}', '{"6":[{"tuplesInWindow":["6"]},{"newTuples":["6"]},{"expiredTuples":["5"]}]}', '{"7":[{"tuplesInWindow":["7"]},{"newTuples":["7"]},{"expiredTuples":["6"]}]}', '{"8":[{"tuplesInWindow":["8"]},{"newTuples":["8"]},{"expiredTuples":["7"]}]}', '{"9":[{"tuplesInWindow":["9"]},{"newTuples":["9"]},{"expiredTuples":["8"]}]}']
[2019-04-18 23:26:22 +0000] [ERROR]: Checking result failed for 20190418231502_IntegrationTest_SlidingTimeWindow1_9c9a09a1-ceaf-4228-a509-b0edd3b876aa topology :: Traceback (most recent call last):
  File "integration_test/src/python/test_runner/main.py", line 204, in run_test
    return results_checker.check_results()
  File "integration_test/src/python/test_runner/main.py", line 123, in check_results
    return self._compare(expected_results, actual_results)
  File "integration_test/src/python/test_runner/main.py", line 136, in _compare
    raise failure
TestFailure: ('Actual result did not match expected result', None)`

**Comment by nwangtw** [comment]
> One of the integration test on a sliding time window bolt has failed the CI. It’s not a stateful topology, isn’t it? I would assume it’s irrelevant to the change in this PR. @nwangtw @joshfischer1108 ideas?
> 
> `[2019-04-18 23:26:22 +0000] [ERROR]: Actual result did not match expected result [2019-04-18 23:26:22 +0000] [INFO]: Actual result ---------- ['{"1":[{"tuplesInWindow":["1"]},{"newTuples":["1"]},{"expiredTuples":[]}]}', '{"2":[{"tuplesInWindow":["2","3"]},{"newTuples":["2","3"]},{"expiredTuples":["1"]}]}', '{"3":[{"tuplesInWindow":["4"]},{"newTuples":["4"]},{"expiredTuples":["2","3"]}]}', '{"4":[{"tuplesInWindow":["5"]},{"newTuples":["5"]},{"expiredTuples":["4"]}]}', '{"5":[{"tuplesInWindow":["6"]},{"newTuples":["6"]},{"expiredTuples":["5"]}]}', '{"6":[{"tuplesInWindow":["7"]},{"newTuples":["7"]},{"expiredTuples":["6"]}]}', '{"7":[{"tuplesInWindow":["8"]},{"newTuples":["8"]},{"expiredTuples":["7"]}]}', '{"8":[{"tuplesInWindow":["9"]},{"newTuples":["9"]},{"expiredTuples":["8"]}]}', '{"9":[{"tuplesInWindow":["10"]},{"newTuples":["10"]},{"expiredTuples":["9"]}]}'] [2019-04-18 23:26:22 +0000] [INFO]: Expected result ---------- ['{"1":[{"tuplesInWindow":["1"]},{"newTuples":["1"]},{"expiredTuples":[]}]}', '{"10":[{"tuplesInWindow":["10"]},{"newTuples":["10"]},{"expiredTuples":["9"]}]}', '{"2":[{"tuplesInWindow":["2"]},{"newTuples":["2"]},{"expiredTuples":["1"]}]}', '{"3":[{"tuplesInWindow":["3"]},{"newTuples":["3"]},{"expiredTuples":["2"]}]}', '{"4":[{"tuplesInWindow":["4"]},{"newTuples":["4"]},{"expiredTuples":["3"]}]}', '{"5":[{"tuplesInWindow":["5"]},{"newTuples":["5"]},{"expiredTuples":["4"]}]}', '{"6":[{"tuplesInWindow":["6"]},{"newTuples":["6"]},{"expiredTuples":["5"]}]}', '{"7":[{"tuplesInWindow":["7"]},{"newTuples":["7"]},{"expiredTuples":["6"]}]}', '{"8":[{"tuplesInWindow":["8"]},{"newTuples":["8"]},{"expiredTuples":["7"]}]}', '{"9":[{"tuplesInWindow":["9"]},{"newTuples":["9"]},{"expiredTuples":["8"]}]}'] [2019-04-18 23:26:22 +0000] [ERROR]: Checking result failed for 20190418231502_IntegrationTest_SlidingTimeWindow1_9c9a09a1-ceaf-4228-a509-b0edd3b876aa topology :: Traceback (most recent call last): File "integration_test/src/python/test_runner/main.py", line 204, in run_test return results_checker.check_results() File "integration_test/src/python/test_runner/main.py", line 123, in check_results return self._compare(expected_results, actual_results) File "integration_test/src/python/test_runner/main.py", line 136, in _compare raise failure TestFailure: ('Actual result did not match expected result', None)`

CI has false alert sometimes. Let me trigger the build again.

**Comment by simingweng** [comment]
@nwangtw CI checks have passed, it's ready for merging I think.

**Pull request merged**